internal/component/mimir/rules/kubernetes/rules.go

```diff
@@ -146,7 +159,60 @@ func New(o component.Options, args Arguments) (*Component, error) {
 	return c, nil
 }
 
+func newNoInit(o component.Options, args Arguments) (*Component, error) {
+	m := newMetrics()
+	m.mustRegister(o.Registerer)
+
+	clusterSvc, err := o.GetServiceData(cluster.ServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("getting cluster service failed: %w", err)
+	}
+
+	c := &Component{
+		log:            o.Logger,
+		opts:           o,
+		args:           args,
+		leader:         newComponentLeadership(o.ID, o.Logger, clusterSvc.(cluster.Cluster)),
+		configUpdates:  make(chan ConfigUpdate),
+		clusterUpdates: make(chan struct{}),
+		ticker:         time.NewTicker(args.SyncInterval),
+		metrics:        m,
+	}
+
+	return c, nil
+}
+
 func (c *Component) Run(ctx context.Context) error {
+	c.startupWithRetries(ctx, c.leader, c, c)
+
+	for {
+		// iteration only returns a sentinel error to indicate shutdown, otherwise it handles
+		// any errors encountered itself by logging and marking the component as unhealthy.
+		if err := c.iteration(ctx, c.leader, c, c); err != nil {
+			break
+		}
+	}
```

Comment:
> Only break if `errors.Is(err, errShutdown)`

I can do this. Would should the behavior for a non-shutdown error be, panic since returning a non-shutdown error is a bug?

Reply:
> I can do this. Would should the behavior for a non-shutdown error be, panic since returning a non-shutdown error is a bug?

I think it can be logged or ignored, potentially combined with setting health. I generally don't want components to have panics because of the blast radius; other components can still continue working even if there's a bug in one of them.

Reply:
Fixed in de9737a21629cbc064f594ef0808125d06dfdb52